### ISSUE TITLE
"Docker Desktop is starting..." forever

### ISSUE THREAD

**Issue by samcro1967** [docker/for-win#5261]
https://hub.docker.com/editions/community/docker-ce-desktop-windows -->
  - [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics - Process does not finish when initiated
  - Diagnostics ID: - Unable to upload.  Diagnostics process starts and never completes.  Log is attached below.

![image](https://user-images.githubusercontent.com/2319445/69994766-3622c900-1514-11ea-9f05-21868c12ea4f.png)

### Expected behavior
Docker Desktop starts successfully

### Actual behavior
Docker Desktop get stuck in starting forever.

### Information
Please, help us understand the problem.  For instance:
  - Is it reproducible? - Yes
  - Is the problem new? - Yes as of about 2 weeks ago
  - Did the problem appear with an update? - Not that I know of
  - Windows Version: 10.0.18363.476
  - Docker Desktop Version:  19.03.5
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM:  Yes

### Steps to reproduce the behavior
Reboot the machine.  It happens almost every time.  All of the containers do start and are accessible.  The Docker API is also functional.

I have uninstalled and reinstalled Docker.  This seems to work for a few days and then the problem reappears.  Occasionally it will start successfully (10% of the time).  If I do a reset, it will work for a while and then stop again.

[log.zip](https://github.com/docker/for-win/files/3913270/log.zip)

**Comment by mikeparker** [comment]
How long did you wait? There is a 10 minute timeout on the boot sequence which will trigger diagnostics and an error report.

**Comment by samcro1967** [comment]
I have waited 24+ hours and it is still stuck in starting.

What other info is needed and how can I get it given that diagnostics is also not responding properly.

**Comment by bguessaoui** [comment]
Hello guys,

Any update?

seeyou

**Comment by briceo** [comment]
I have also encountered this issue.  The 10 minute time out doesn't seem to work.  I even tried switching from Linux containers to Windows containers, and it just hangs forever.

Windows Version: 10.0.19041.84 (using WSL2 features)
Docker Engine: 19.03.5
Docker Desktop: 2.2.0.3 (42716)

In my case, it was working fine yesterday, but the battery on my laptop died abruptly, and turned off my system.  I turned it back on the next day, and Docker started doing the hang forever on startup.  The rest of the system looks fine.  The Docker daemon is not running on either windows or linux.

Any thing I can look at or try?

**Comment by reidj32** [comment]
I have also encountered this issue. I was already using WSL2 without issue, and decided to upgrade to Docker Desktop Edge to utilize the new WSL2 integration. I installed Docker Desktop Edge without issue and started using Linux containers right away in the WSL2 environment. The trouble is with a machine restart. Once the machine is restarted, docker will hang on startup forever, and never hit the 10 minute timeout. The only resolution I have found is to completely uninstall Docker Desktop, remove all data directories (AppData, ProgramData, etc) and reinstall. I am also using Windows containers (supporting legacy systems), and I am unable to switch between them once this issue starts to occur. I have also tried another suggest workaround where I would call the following commands from an elevated prompt, but it had no effect for me.

`net stop com.docker.service`
`net start com.docker.service`

So for now, I have to reinstall docker completely with each restart, and then download all my images again (very slow for my Windows containers).

Windows Version: 10.0.19041.84
Docker Desktop Version: 2.2.3.0 (43965)
Diagnostics ID: 77A64FB6-9C39-4E86-B25F-14BBE5EFC5FA/20200415172857

Something I noticed in the logs, the system appears to be trying to communicate with a server located at localhost:80, I am running a local IIS server already listening on port 80 and 443. Could this be part of the issue?

**Comment by valfrid-ly** [comment]
I just installed last stable version of docker-desktop and ran into the same problem.

Got here trying to find a solution

**Comment by felubra** [comment]
I was facing the same issue. In my case the firewall was blocking connections made by the binaries:
- `C:\program files\docker\docker\resources\vpnkit.exe`
- `C:\program files\docker\docker\com.docker.service`
- `C:\program files\docker\docker\resources\com.docker.proxy.exe`
After allowing this applications, I was able to start.

**Comment by valfrid-ly** [comment]
I just installed the last update and allowed "vpnkit" and "docker.proxy" in the firewall. It's working!! "docker.service" doesn't exists in my system

**Comment by lordvlad** [comment]
Adding vpnkit and docker.proxy binaries to the firewall did not make a difference. Removing other WSL distros did not make a difference either.
The frontend shows `Docker updating` in the lower left-hand corner, and does not react to any input. Docker process seems very idle at < 1%CPU. 
I can't uninstall using the regular windows uninstall process, had to forcefully purge from disk and registry, reinstalled, all to no avail.
Installer says its version 2.3.0.5

@wemrysi what process are you killing? Can you provide a step-by-step?

**Comment by wemrysi** [comment]
@lordvlad I mentioned it in the previous post, in my case the process was

```
zsh -c mkdir -p ~/.docker
```

More generally, it looks like Docker Desktop executes some commands using the default login shell on the WSL2 instance. My zsh configuration runs a utility that requires user input the first time it is run. If Docker Desktop is the first access to the WSL2 instance, it gets stuck waiting for input due to my aforementioned configuration.

I've managed to work around this issue by keeping `bash` as my login shell with default configuration and running `zsh` (with all of my custom initialization) manually. The result is that Docker Desktop is no longer impeded due to awaiting input and is able to start successfully.

TL;DR if you have anything in your login shell configuration that requires user input to proceed, it may be preventing Docker Desktop from starting.

I couldn't find much documentation about what commands, etc. are run by Docker Desktop in the default WSL2 instance. If it isn't doing so already, it might be useful if Docker Desktop defined a unique environment variable that is set when running commands (similar to, say, CI platforms), so that users could condition on it to avoid this problem.

**Comment by lordvlad** [comment]
@wemrysi Thanks for the lengthy explanation. As I mentioned before, I even removed all other WSL distros to no avail.

In the meantime, I had two more tries at cleaning up:
- registry
- appdata/{roaming,local}
- programdata
- userprofile/.docker

and installing the stable version. In the end it worked. Not sure  which step exactly fixed it, but now docker works as its supposed to.

**Comment by HyperCharlie** [comment]
Having this occur as well. Docker is actually running (I can use it from the command line), but in the Dashboard app it says "Starting" forever, and does not display running containers.

**Comment by elitmiar** [comment]
Had the same issue as @valfrid-ly and after allowing the below binaries in the Windows Defender Firewall as a Inbound Rule, Docker started up successfully.

```
%ProgramFiles%\Docker\Docker\resources\com.docker.proxy.exe
%ProgramFiles%\Docker\Docker\resources\vpnkit.exe
```
